src/backend/renderer/mod.rs

```diff
@@ -410,12 +480,77 @@ impl<R: Renderer + ImportShm + ImportDma> ImportAll for R {
     ) -> Option<Result<<Self as Renderer>::TextureId, <Self as Renderer>::Error>> {
         match buffer_type(buffer) {
             Some(BufferType::Shm) => Some(self.import_shm_buffer(buffer, surface, damage)),
-            Some(BufferType::Dma) => Some(self.import_dma_buffer(buffer)),
+            Some(BufferType::Dma) => Some(self.import_dma_buffer(buffer, surface, damage)),
             _ => None,
         }
     }
 }
 
+/// Trait for renderers supporting exporting contents of framebuffers or textures into memory.
+pub trait ExportMem: Renderer {
+    /// Texture type representing a downloaded pixel buffer.
+    type TextureMapping: TextureMapping;
+
+    /// Copies the contents of the currently bound framebuffer.
+    ///
+    /// This operation is not destructive, the contexts of the framebuffer keep being valid.
+    ///
+    /// This function *may* fail, if (but not limited to):
+    /// - The framebuffer is not readable
+    /// - The region is out of bounds of the framebuffer
+    /// - There is not enough space to create the mapping
+    fn copy_framebuffer(
```

Comment:
These copy_framebuffer functions must create a new `Renderer::Mapping`.

Is this a correct analysis of the docs here?

Reply:
yes, a single mapping may be re-used to read the same contents multiple times, but each call of `copy_` creates a new "copy".